cob/cob.c

```diff
@@ -352,6 +352,154 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
 	.ko_compare = ns_cmp
 };
 
+/**
+ * Returns struct m0_cob_bckey size.
+ */
+static size_t m0_cob_bckey_size(void)
+{
+	return sizeof(struct m0_cob_bckey);
+}
+
+/**
+ * Returns struct m0_cob_bcrec size.
+ */
+static size_t m0_cob_bcrec_size(void)
+{
+	return sizeof(struct m0_cob_bcrec);
+}
+
+static m0_bcount_t bc_ksize(const void *key)
+{
+	return sizeof(struct m0_cob_bckey);
+}
+
+static m0_bcount_t bc_vsize(const void *val)
+{
+	return sizeof(struct m0_cob_bcrec);
+}
+
+/**
+ * Make bytecount key for iterator. Allocate space for key.
+ */
+static int m0_cob_bckey_make(struct m0_cob_bckey **keyh,
+				       const struct m0_fid  *pver_fid,
+				       uint64_t              user_id)
+{
+	struct m0_cob_bckey *key;
+
+	key = m0_alloc(m0_cob_bckey_size());
+	if (key == NULL)
+		return M0_ERR(-ENOMEM);
+	key->cbk_pfid = *pver_fid;
+	key->cbk_user_id = user_id;
+	*keyh = key;
+	return 0;
+}
+
+M0_INTERNAL int m0_cob_bckey_cmp(const struct m0_cob_bckey *k0,
+				 const struct m0_cob_bckey *k1)
+{
+	int rc;
+
+	M0_PRE(m0_fid_is_set(&k0->cbk_pfid));
+	M0_PRE(m0_fid_is_set(&k1->cbk_pfid));
+
+	rc = m0_fid_cmp(&k0->cbk_pfid, &k1->cbk_pfid);
+	return rc ?: k0->cbk_user_id == k1->cbk_user_id ? 0 : -EINVAL;
+}
+
+/**
+ * Bytecount table comparator.
+ */
+static int bc_cmp(const void *key0, const void *key1)
+{
+	return m0_cob_bckey_cmp((const struct m0_cob_bckey *)key0,
+				(const struct m0_cob_bckey *)key1);
+}
+
+M0_INTERNAL int m0_cob_bc_iterator_init(struct m0_cob             *cob,
+					struct m0_cob_bc_iterator *it,
+					const struct m0_fid       *pver_fid,
+					uint64_t                   user_id)
+{
+	int rc;
+
+	/* Prepare entry key using passed started position. */
+	rc = m0_cob_bckey_make(&it->ci_key, pver_fid, user_id);
+	if (rc != 0)
+		return M0_RC(rc);
+
+	it->ci_rec = m0_alloc(m0_cob_bcrec_size());
+	if (it->ci_rec == NULL) {
+		m0_free(it->ci_key);
+		return M0_RC(rc);
+	}
+
+	m0_be_btree_cursor_init(&it->ci_cursor, &cob->co_dom->cd_bytecount);
+	it->ci_cob = cob;
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_cob_bc_iterator_get(struct m0_cob_bc_iterator *it)
+{
+	struct m0_cob_bckey *bckey;
+	struct m0_cob_bcrec *bcrec;
+	struct m0_buf        key;
+	struct m0_buf        val;
+	int                  rc;
+
+	m0_buf_init(&key, it->ci_key, m0_cob_bckey_size());
+	m0_buf_init(&val, it->ci_rec, m0_cob_bcrec_size());
+	rc = m0_be_btree_cursor_get_sync(&it->ci_cursor, &key, true);
+	if (rc == 0) {
+		m0_be_btree_cursor_kv_get(&it->ci_cursor, &key, &val);
+		bckey = (struct m0_cob_bckey *)key.b_addr;
+		bcrec = (struct m0_cob_bcrec *)val.b_addr;
+
+		M0_ASSERT(sizeof(bckey) <= m0_cob_bckey_size());
+		M0_ASSERT(sizeof(bcrec) <= m0_cob_bcrec_size());
+		memcpy(it->ci_key, bckey, m0_cob_bckey_size());
+		memcpy(it->ci_rec, bcrec, m0_cob_bcrec_size());
+	}
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_cob_bc_iterator_next(struct m0_cob_bc_iterator *it)
```

Comment:
[Suggestion] Check if iterator_next and iterator_get can be combined into a single API, with an additional argument to select next operation.

Reply:
m0_cob_bc_iterator_get() operation we expect iterator already initialized and in m0_cob_bc_iterator_next(), we move the cursor position by one. By looking other implementation in file for iterator functions, I think keeping this as it is would be correct fit.